pulumi/infra/firecracker_assets.py

```diff
@@ -0,0 +1,100 @@
+from pathlib import Path
+from typing import Optional
+
+import pulumi_aws as aws
+from infra import config
+from infra.artifacts import ArtifactGetter
+from infra.bucket import Bucket
+from infra.path import path_from_root
+
+import pulumi
+
+FIRECRACKER_KERNEL_FILENAME = "firecracker_kernel.tar.gz"
+
+
+class FirecrackerAssets(pulumi.ComponentResource):
+    """
+    Uploads Firecracker assets from disk-or-Cloudsmith into
+    S3 bucket.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        repository_name: Optional[str],
+        artifacts: ArtifactGetter,
+        opts: pulumi.ResourceOptions = None,
+    ) -> None:
+        super().__init__("grapl:FirecrackerAssets", name, None, opts)
+
+        self.kernel_asset = local_or_remote_asset(
+            local_path=path_from_root("dist/firecracker_kernel.tar.gz"),
+            artifacts=artifacts,
+            artifact_key=FIRECRACKER_KERNEL_FILENAME,
+            repository_name=repository_name,
+        )
+
+
+class FirecrackerS3BucketObjects(pulumi.ComponentResource):
+    def __init__(
+        self,
+        name: str,
+        plugins_bucket: Bucket,
+        firecracker_assets: FirecrackerAssets,
+        opts: pulumi.ResourceOptions = None,
+    ) -> None:
+        super().__init__("grapl:FirecrackerS3BucketObjects", name, None, opts)
+        child_opts = pulumi.ResourceOptions(parent=self)
+
+        kernel_s3obj = aws.s3.BucketObject(
+            "firecracker_kernel",
+            key=FIRECRACKER_KERNEL_FILENAME,
+            bucket=plugins_bucket.bucket,
+            source=firecracker_assets.kernel_asset,
+            opts=child_opts,
+        )
```

Comment:
I think this is going to effectively mean that we'll only ever have a single kernel image in use at a time, which may not be a good idea.

Also, since the default is to create a new resource and then delete the old resource, this will also have the effect of deleting the kernel image from the bucket entirely once we get a second kernel version in play (_"Ask me how I know", he said bitterly_). At the very least, we'll need to have a `delete_before_replace=True` resource option.

However, this this is also why I think we'll want to be able to have multiple kernel images in the bucket at a time; imagine a Nomad job trying to spin up right between the deletion of the old image and the uploading of the new image. It won't find any image, and wackiness will ensue.

Reply:
> since the default is to create a new resource and then delete the old resource
I'm not sure how that works with S3BucketObject here, since - _theoretically_ - an update would just be an S3 PUT request that overwrites the object already there, no?

Some other thoughts:
- Do you mean `delete_before_replace=False`? Because what we want to avoid is having _nothing_ at that key, right?
- It looks like Nomad doesn't cache its artifacts, what the heck?! we gotta download another 20mb from s3 every time we run a plugin? https://github.com/hashicorp/nomad/issues/5217

> this is also why I think we'll want to be able to have multiple kernel images in the bucket at a time
- You've made a good case for why we'd want it
- I don't know how to do this today with our existing artifacts setup
- I'm open to ideas
- I'd ask if this is something that needs to be solved ASAP, or could be kicked down the road

Reply:
I mean `delete_before_replace=True`, because the artifacts are going to be _at the same location_ in the bucket.

If we have `delete_before_replace=False`, then this is what will happen:
- We upload the new kernel to `<BUCKET>/firecracker_kernel` (the "replace" step)
- Now that the new resource is in place, we can delete the old one; now we delete `<BUCKET>/firecracker_kernel`
- Now we have no kernel

If we have `delete_before_replace=True`, this is what happens:
- Delete the old resource `<BUCKET>/firecracker_kernel`
- Upload the new resource to `<BUCKET>/firecracker_kernel`
- Now we have a new kernel.

The latter is preferable to the former, but it's still not perfect because there is still a time where there is no kernel available; anything that spins up in those few seconds is going to have a bad time.

For now, I think this is fine (let's get to the point where we're actually running things in Firecracker first, you know 😅), but we will definitely need to address this before production. An issue to track this would be great.

(Incidentally, I don't think it should be too terribly hard, or significantly different than what we already do with our Cloudsmith artifacts... we just incorporate the version into the key, so rather than downloading from `<BUCKET>/firecracker_kernel`, we go to `<BUCKET>/firecracker_kernel/kernel_x.y.z`... probably good to also include some metadata files alongside that -- checksums, etc. But yes, all this could come after this PR. We used to use S3 as a glorified artifact store like this at my last job.)

(Is the reason we're putting these in S3 rather than Cloudsmith data transfer fees? Or do we know enough about that to make a call at this point?)

Reply:
Okay, that makes sense.

S3-vs-Cloudsmith-wise: I mostly went this route because the plugin.nomad file provided in the spec said
```

      artifact {
        source      = var.kernel_artifact_url
        destination = "local/vmlinux"
        headers {
          x-amz-expected-bucket-owner = var.aws_account_id
          x-amz-meta-client-id        = "nomad-deployer"
        }
      }
```
and I was blindly following it instead of questioning 😇 
but this does suggest that we want to have some acl checks on it

Reply:
🤔 If that's just a URL with HTTP headers, we should be able to use Cloudsmith just as easily.

(Just food for thought, and possibly for the follow-up issue.)